app/models/concerns/second_factor_manager.rb

```diff
@@ -23,18 +24,88 @@ def authenticate_totp(token)
     totp = self.totp
     last_used = 0
 
-    if self.user_second_factor.last_used
-      last_used = self.user_second_factor.last_used.to_i
+    if self.user_second_factors.totp.last_used
+      last_used = self.user_second_factors.totp.last_used.to_i
     end
 
     authenticated = !token.blank? && totp.verify_with_drift_and_prior(token, 30, last_used)
-    self.user_second_factor.update!(last_used: DateTime.now) if authenticated
+    self.user_second_factors.totp.update!(last_used: DateTime.now) if authenticated
     !!authenticated
   end
 
   def totp_enabled?
-    !!(self&.user_second_factor&.enabled?) &&
+    !!(self&.user_second_factors&.totp&.enabled?) &&
       !SiteSetting.enable_sso &&
       SiteSetting.enable_local_logins
   end
+
+  def backup_codes_enabled?
+    !!(self&.user_second_factors&.backup_codes&.present?) &&
+      !SiteSetting.enable_sso &&
+      SiteSetting.enable_local_logins
+  end
+
+  def authenticate_second_factor(token, second_factor_method)
+    if second_factor_method == UserSecondFactor.methods[:totp]
+      authenticate_totp(token)
+    elsif second_factor_method == UserSecondFactor.methods[:backup_codes]
+      authenticate_backup_code(token)
+    end
+  end
+
+  def generate_backup_codes
+    codes = []
+    10.times do
+      codes << SecureRandom.hex(8)
+    end
+
+    salt = SecureRandom.hex(16)
+    codes_json = codes.map do |code|
+      { :salt => salt,
```

Comment:
Trivial but we prefer the `salt: salt, code_hash: etc` syntax. I'm not sure why rubocop didn't mention this!